### PR TITLE
feat: Add header file for generic binary search

### PR DESCRIPTION
**Pull request by Albirair**
Add the header file (search/binary_search.h) which defines the function template "binary_search" in 2 forms, and the source file (search/binary_search_test.cpp) for testing the functions.

### CODE REVIEW

**Comment by kvedala** [comment]
Excellent documentation and coding standard 👍

File: search/binary_search_test.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/// Copyright 2020 @author Albirair
```
Comment (kvedala) [inline search/binary_search_test.cpp:1]: could you also add
```c++
/**
  * @file
  *
  * A generic implementation of binary_search.
  */
```
to this file for consistency. I am building an document generator [here (repo)](https://github.com/kvedala/C-Plus-Plus) that creates the [online documentation](https://kvedala.github.io/C-Plus-Plus)

Comment (Albirair) [reply search/binary_search_test.cpp]: I added the **file** Doxygen command to `search/binary_search_test.cpp` as suggested by @kvedala in this [latest commit.](https://github.com/TheAlgorithms/C-Plus-Plus/pull/738/commits/16fec03ac899cdc847f20e553de7850a83caaa71)

File: search/binary_search_test.cpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * @file
+ *
+ * A generic implementation of binary_search.
+ */
+#include "./binary_search.h"
+#include <iostream>
+#include <vector>
+#include <random>
+#include <algorithm>
+int main(int argc, char ** argv) {
```
Comment (kvedala) [inline search/binary_search_test.cpp:12]:
```suggestion

/** Main function */
int main(int argc, char ** argv) {
```

File: search/binary_search_test.cpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    using std::size_t, std::cout, std::cin;
+    std::default_random_engine e;
+    size_t n;
+    do {
```
Comment (kvedala) [inline search/binary_search_test.cpp:16]:
```suggestion

    do {
```

File: search/binary_search_test.cpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        cout << "enter a nonzero number of elements\n";
+        cin >> n;
+    } while (!n);
+    std::vector<int> v(n);
```
Comment (kvedala) [inline search/binary_search_test.cpp:20]:
```suggestion

    std::vector<int> v(n);
```

File: search/binary_search_test.cpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    int val;
+    for (size_t i(0); i != n; ++i)
+        v[i] = e() % 100;
+    std::sort(v.begin(), v.end());
```
Comment (kvedala) [inline search/binary_search_test.cpp:24]:
```suggestion

    std::sort(v.begin(), v.end());
```

File: search/binary_search_test.cpp [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    for (size_t i(0); i != n; ++i)
+        cout << "v[" << i << "] = " << v[i] << '\n';
+    cout << "enter the value to search for:\t";
```
Comment (kvedala) [inline search/binary_search_test.cpp:27]:
```suggestion

    cout << "enter the value to search for:\t";
```

File: search/binary_search_test.cpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    cin >> val;
+    auto r(binary_search(v.begin(), n, val));
+    if (r.first)
```
Comment (kvedala) [inline search/binary_search_test.cpp:30]:
```suggestion

    if (r.first)
```

File: search/binary_search_test.cpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        cout << val << " is at " << r.second - v.begin();
+    else
+        cout << val << " doesn't exist";
+    return 0;
```
Comment (kvedala) [inline search/binary_search_test.cpp:34]:
```suggestion

    return 0;
```

**Pull request closed without merging**